moc_openshift.py

```diff
@@ -0,0 +1,471 @@
+                            }
+                        ),
+                        status=400,
+                        mimetype="application/json",
+                    )
+                role_binding["userNames"].remove(user)
+            else:
+                self.logger.info("Error: unknown update operation")
+                return Response(
+                    response=json.dumps(
+                        {
+                            "msg": f"Invalid request ({user},{project_name},role,{operation})"
+                        }
+                    ),
+                    status=400,
+                    mimetype="application/json",
+                )
+
+            # now add or remove the user
+            result = self.update_rolebindings(
+                project_name, openshift_role, role_binding
+            )
+
+            msg = "unknown message"
+            if result.status_code in (200, 201):
+                if operation == "add":
+                    msg = "Added role to user on project"
+                elif operation == "del":
+                    msg = "removed role from user on project"
+                return Response(
+                    response=json.dumps({"msg": msg}),
+                    status=200,
+                    mimetype="application/json",
+                )
+            if operation == "add":
+                msg = "unable to add role to user on project"
+            elif operation == "del":
+                msg = "unable to remove role from user on project"
+            return Response(
+                response=json.dumps({"msg": msg}),
+                status=400,
+                mimetype="application/json",
+            )
+        return Response(
+            response=json.dumps(
+                {"msg": f"rolebinding already exists ({user},{project_name},{role})"}
+            ),
+            status=400,
+            mimetype="application/json",
+        )
+
+
+class MocOpenShift4x(MocOpenShift):
+    """API implementation for OpenShift 4.x"""
+
+    # member functions for projects
+    def project_exists(self, project_name):
+        url = f"{self.get_url()}/apis/project.openshift.io/v1/projects/{project_name}"
+        result = self.get_request(url, True)
+        if result.status_code in (200, 201):
+            return True
+        return False
+
+    def create_project(self, project_name, display_name, user_name):
+        # check project_name
+        url = f"{self.get_url()}/apis/project.openshift.io/v1/projects/"
+        payload = {
+            "kind": "Project",
+            "apiVersion": "project.openshift.io/v1",
+            "metadata": {
+                "name": project_name,
+                "annotations": {
+                    "openshift.io/display-name": display_name,
+                    "openshift.io/requester": user_name,
+                },
+            },
+        }
+        return self.post_request(url, payload, True)
+
+    def delete_project(self, project_name):
```

Comment:
We should be performing some safety checks here. With this implementation, we would permit a request such as `DELETE /projects/openshift-apiserver`. One option (that we use in https://github.com/CCI-MOC/moc-project-tool) is to label all projects created by this service with a specific label (e.g., `massopen.cloud/project`), and then only operate on projects that have that label.